### ISSUE TITLE
Web Console not displaying events after migrating to v2

### ISSUE THREAD

**Issue by ClementDBS** [ServerlessLife/serverless-spy#57]
I often use the serverless-spy web console to debug my app. This worked perfectly in v1.

However, after migrating to v2, the web console no longer displays any events when I run my tests. The console itself is running correctly, and I can see connection logs on my terminal, but no events are shown on the page. There are no errors or warnings in the logs that might indicate what is causing the issue.

I'm using nodejs 20 for my lambdas and serverless-spy 2.3.3.
Are there any new configurations or changes needed to ensure the console receives and displays events in v2 ? Can you verify if this issue is reproducible on your end ?

**Comment by Lewenhaupt** [comment]
Hi @ClementDBS,

I should have some time to look into this tomorrow, but in the meantime, are you able to write tests using v2? The major thing we changed between v1 and v2 was to start using AWS IoT. What I know can happen when activating AWS IoT the first time in an account, is that it can take a couple of minutes before it actually works. But that should then impact both the console and the listener for the tests.

**Comment by Lewenhaupt** [comment]
@ClementDBS I think I'm able to reproduce it. Could you check if the ws-connection is receiving messages if you check the network tab in the browsers devtools? I can see messages are being received but nothing is displayed on my end.

**Comment by Lewenhaupt** [comment]
@ClementDBS @markostru I'm pretty sure I understand the issue. We seem to have an issue with how we set the topic in the actual AWS environment and how the cli tries to translate stacks to topics because the format in the cdkoutput file. In the interceptors we use `SSPY_ROOT_STACK` which is a cleaned up version of the root stacks id. This isn't the same as the key in the output. 
@markostru I think one way we could solve this is to add the SSPY_ROOT_STACK variable to the cdkoutput file under each stack such that we can resolve the true value based on the selected stack, this way we could also start subscribing the the correct topic instead of the full wildcard topic.

**Comment by ServerlessLife** [comment]
I am sorry I did not have time to respond sooner. 🎄🎅

@Lewenhaupt 
It took me a bit of time to understand the issue. Kudos to you, @Lewenhaupt, for figuring it out so quickly. 

The reason we have `SSPY_ROOT_STACK` is to uniquely identify the project and use it as an IoT topic. We could use some other identification instead of the root stack name. For Lambda Live Debugger, I use [a hashed value of a combination of the project folder and machine ID](https://github.com/ServerlessLife/lambda-live-debugger/blob/main/src/configuration.ts#L80). Maybe that could simplify things? WDYT? :thinking:

**Comment by Lewenhaupt** [comment]
@markostru Yeah that would be a good refactoring, but a bit of a bigger one I think? And we would still need to pass some mapping to the cli from the cdkOutput as we'd probably pass the "debuggerID" to the environment of the lambdas right? 
Have a look at the draft PR i created (I wasn't able to test i yet as my poor personal laptop runs out of ram if I try to run the tests and the cli at the same time :joy:). I simply output the SSPY_ROOT_STACK value in the cdkOutput such that the cli can build a mapping between stackName and the topic we use and the web console can then use that mapping when doing the filtering. I also disabled filtering altogether if the stackList wasn't available/mappings weren't available.

Let me know what you think :slightly_smiling_face: and if you're able to help me test it that would be great, otherwise I can probably test it sometime tomorrow on my work laptop.

**Comment by ServerlessLife** [comment]
@ClementDBS 
Can you confirm that you are still having an issue? As @Lewenhaupt mentioned, ServerlessSpy V2 uses AWS IoT, and it can take a few minutes to become available when you first start using it. That is a one-time issue that does not reoccur after that.

-------------------------------
@Lewenhaupt 
Please ignore my suggestion for `debuggerID`. I did not know/remember that we were using a wildcard.  `debuggerID` would be a way to avoid that. But yes, it would require passing the value as the environment available. That would be a good suggestion for the next refactoring.

I apologize for the following basic questions. :pray: :pray: :pray:  I have not worked on the code much for some time, and most of these things were written by you.

> **_In the interceptors we use SSPY_ROOT_STACK which is a cleaned up version of the root stacks id. This isn't the same as the key in the output._**

What do you mean by being different?
Environment variable is set like this:
```typescript
    func.addEnvironment(
      envVariableNames.SSPY_ROOT_STACK,
      this.cleanName(this.findRootStack(Stack.of(this)).node.id)
    );
```

And the output is the same.

```typescript
    const rootStack = this.cleanName(
      this.findRootStack(Stack.of(this)).node.id
    );

  new CfnOutput(this, 'ServerlessSpyIoTEndpoint', {
      key: 'ServerlessSpyWsUrl',
      value: `${this.iotEndpoint}/${rootStack}`,
    });
```    
The only difference is that it has an additional IoT endpoint.

Consequently, I am also not sure why you would put [that to a new output](https://github.com/ServerlessLife/serverless-spy/pull/60/files#diff-54ceebbba7d1f9c1b8f883627e43918eb421aaa88d377e573ede489bbe14f3a2R107) and add an additional mapping [here](https://github.com/ServerlessLife/serverless-spy/pull/60/files#diff-e875f7ba15c18cc6e6ec4dddb5c928142558cf8680d660b0c83ef61b114ef269R135).

**Comment by ClementDBS** [comment]
Hi @Lewenhaupt @ServerlessLife, sorry for the late answer and thank you for your reactivity !

I deployed my stack with the new version, waited for 30minutes and I still have the issue on my side. I have nothing on my browser network tab either.
Am i supposed to have a ServerlessSpyTopic or SSPY_ROOT_STACK in my cdkOuput ? I only have my own variables and ServerlessSpyWsUrl.

**Comment by ServerlessLife** [comment]
Hi @ClementDBS 
Some more questions:
1. Do you have some integration tests? Do they work?
2. Can you tell us more about your setup? Do you use multiple stacks? So
3. Which parameters do you use when you run the Web Console? Can you paste the whole CLI command?

**Comment by ClementDBS** [comment]
Hi @ServerlessLife,

Thanks for your follow-up. Here's the information you requested:

1. Integration Tests
I only run end-to-end tests. In these tests, I publish events to EventBridge and use serverless-spy to wait for the new events that my stack emits in response. The tests themselves work fine, just as they did in v1.

2. Setup
I am testing a single stack here. I'm deploying the stack to a test environment and running the tests on it.

3. CLI Command
Here’s the CLI command I use to run the web console:
``ENV AWS_PROFILE=perso npx sspy --cdkoutput test.output.json``
The test.output.json file is the CDK output generated for the stack.

Let me know if you need additional details!

**Comment by ServerlessLife** [comment]
@ClementDBS 
Thank you for the details.
That is really strange and I do not see any reason not to work. No `SSPY_ROOT_STACK` variable on anything else is needed. That is an internal ServerlessSpy Lambda environment variable.

Can you test the sample project: https://github.com/ServerlessLife/serverless-spy-example
Or can you provide a sample project that does not work?

**Comment by Lewenhaupt** [comment]
@markostru the issue is that the stackList in the cli is populated based on the keys in the cdkOuput which do not have the same format as the output of cleanName(), which means that when the cli tries to use the key from cdkOuput to filter the list that key is not the same as the topic set on the message. See the changes I did on the other branch.
@ClementDBS if you are able to build the branch yourself you can try that version. I might have time later today to test it myself but not a 100% sure.

**Comment by ClementDBS** [comment]
My stack name is "test-billing-line-control." I have multiple microservices, and I’ve tried using the web console on several of them, but none are working. There aren’t any special characters in the stack names.

I’ll let you work on the fix, and I’ll keep you updated if the problem persists on my side afterward. I’ll also try to investigate more when I have time.

Thanks again for your help! 🙏

**Comment by ServerlessLife** [comment]
@ClementDBS 
The "-" character in the stack name is the problem. Strange, that nobody noticed it before.

**Comment by Lewenhaupt** [comment]
> @ClementDBS The "-" character in the stack name is the problem. Strange, that nobody noticed it before.

It's actually possible for more problems to arise since the id of the node in the stack is not the same as the stack name. Usually people use the same or autogenerate one, but I set both explicitly.

I've tested my branch so I've moved it from draft to open PR and requested your review @markostru 

> @Lewenhaupt I finally understand 🙌 . But you do not have to add a new output. The `rootStack` is now duplicated, once with an additional endpoint, which can easily be parsed out.
> 
> ```ts
>     new CfnOutput(this, 'ServerlessSpyIoTEndpoint', {
>       key: 'ServerlessSpyWsUrl',
>       value: `${this.iotEndpoint}/${rootStack}`,
>     });
>     new CfnOutput(this, 'ServerlessSpyTopic', {
>       key: 'ServerlessSpyTopic',
>       value: `${rootStack}`,
>     });
> ```
> 
> You also did not implement (or forgot to commit) the endpoint `/stackTopicMappings`. So this can not work.
> 
> @ClementDBS TL;DR The most likely reason is that your stack name contains a special character. We will fix that.

Yeah we could do it that way but that would break it if we changed the format of the IoT topic used and didn't update the message. I think this would be more reliable?

**Comment by ServerlessLife** [comment]
@ClementDBS 
I merged & deployed & tested the fix that @Lewenhaupt created.

Thank you @Lewenhaupt very much for the fix.

**Comment by Lewenhaupt** [comment]
> @ClementDBS 
> I merged & deployed & tested the fix that @Lewenhaupt created.
> 
> Thank you @Lewenhaupt very much for the fix. 

And thanks for a good review @markostru 😊
